### PR TITLE
Building-Blocks-Enumerables - Ruby

### PR DESCRIPTION
**Pull request by gasb150**
This project involves creating some commonly used enumerable methods (#my_each, #my_each_with _index, #my_select, #my_all, #my_any?, #my_none, #my_count, #my_map, #my_inject), stored on a script file, the new methods are onto the existing Enumerable module, #my_inject was tested creating a method called #multiply_els.

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    end
+    self
+  end
+
+  def my_select
+    arr = self
+    n_array = []
+    arr.my_each { |x| n_array << x if yield x }
+    n_array
+  end
```
Comment (PhillipUg) [inline enumerables.rb]: - [ ] These three methods should return an `Enumerator` if no block is given

**Review by PhillipUg — CHANGES_REQUESTED** [review]
Hey team, 👋

You have done an amazing job implementing the project requirements of this project. Your `my_inject` and `multiply_els` methods work very well.

However, there are still some changes 👇 you need to implement to approve this request 🔍

### Changes Required ♻️ 

- [ ] Your `my_all?`, `my_any?`, `my_none?`, `my_count?` and `my_map` need more work. So i am going to provide you with some custom test scenarios. Please make sure your methods pass all of them before submitting for re-review.

```

# 4. my_all? (example test cases)
puts 'my_all?'
puts '-------'
p [3, 5, 7, 11].my_all?(&:odd?) # => true
p [-8, -9, -6].my_all? { |n| n < 0 } # => true
p [3, 5, 8, 11].my_all?(&:odd?) # => false
p [-8, -9, -6, 0].my_all? { |n| n < 0 } # => false
# test cases required by tse reviewer
p [1, 2, 3, 4, 5].my_all? # => true
p [1, 2, 3].my_all?(Integer) # => true
p %w[dog door rod blade].my_all?(/d/) # => true
p [1, 1, 1].my_all?(1) # => true
puts

# 5. my_any? (example test cases)
puts 'my_any?'
puts '-------'
p [7, 10, 3, 5].my_any?(&:even?) # => true
p [7, 10, 4, 5].my_any?(&:even?) # => true
p %w[q r s i].my_any? { |char| 'aeiou'.include?(char) } # => true
p [7, 11, 3, 5].my_any?(&:even?) # => false
p %w[q r s t].my_any? { |char| 'aeiou'.include?(char) } # => false
# test cases required by tse reviewer
p [3, 5, 4, 11].my_any? # => true
p [1, nil, false].my_any?(1) # => true
p [1, nil, false].my_any?(Integer) # => true
p %w[dog door rod blade].my_any?(/z/) # => false
p [1, 2, 3].my_any?(1) # => true
puts

# 6. my_none? (example test cases)
puts 'my_none?'
puts '--------'
p [3, 5, 7, 11].my_none?(&:even?) # => true
p %w[sushi pizza burrito].my_none? { |word| word[0] == 'a' } # => true
p [3, 5, 4, 7, 11].my_none?(&:even?) # => false
p %w[asparagus sushi pizza apple burrito].my_none? { |word| word[0] == 'a' } # => false
# test cases required by tse reviewer
p [1, 2, 3].my_none? # => false
p [1, 2, 3].my_none?(String) # => true
p [1, 2, 3, 4, 5].my_none?(2) # => false
p [1, 2, 3].my_none?(4) # => true
puts

# 7. my_count (example test cases)
puts 'my_count'
puts '--------'
p [1, 4, 3, 8].my_count(&:even?) # => 2
p %w[DANIEL JIA KRITI dave].my_count { |s| s == s.upcase } # => 3
p %w[daniel jia kriti dave].my_count { |s| s == s.upcase } # => 0
# test cases required by tse reviewer
p [1, 2, 3].my_count # => 3
p [1, 1, 1, 2, 3].my_count(1) # => 3
puts

# 8. my_map
puts 'my_map'
puts '------'
p [1, 2, 3].my_map { |n| 2 * n } # => [2,4,6]
p %w[Hey Jude].my_map { |word| word + '?' } # => ["Hey?", "Jude?"]
p [false, true].my_map(&:!) # => [true, false]
my_proc = proc { |num| num > 10 }
p [18, 22, 5, 6].my_map(my_proc) { |num| num < 10 } # => true true false false
puts
```

#### Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack as **Phillip Musiime**
I would be happy to help! 👩‍💻 

Happy coding 🤓

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      return true
+    elsif arg.is_a?(Class)
+      arr.my_each { |x| return false unless x.is_a?(arg) }
+    elsif arg.is_a?(Regexp)
+      arr.my_each { |x| return false unless arg.match(x) }
+    else
+      arr.my_each { |x| return false if x != arg }
+    end
+    y
+  end
```
Comment (BrittanyBlake) [inline enumerables.rb]: - [ ] When no block or argument is given, `#my_all?` should return false when one of the collection members are false or nil.

```ruby
 [nil, false, true, []].all? === [nil, false, true, []].my_all? => false
```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      arr.my_each { return true unless arr.nil? }
+    elsif arg.is_a?(Class)
+      arr.my_each { |x| return true if x.is_a?(arg) }
+    elsif arg.is_a?(Regexp)
+      arr.my_each { |x| return true if arg.match(x) }
+    else
+      arr.my_each { |x| return true if x == arg }
+    end
+    false
+  end
```
Comment (BrittanyBlake) [inline enumerables.rb]: - [ ] When no block or argument is given, `#my_any?` should return false if at least one of the members of the collection is not true.

```ruby
[nil, false, nil, false].any? === [nil, false, nil, false].my_any? => false
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+  def my_none?(arg = nil)
+    arr = self
+    return !arr.my_any?(arg) unless block_given?
+
+    arr.my_each { |x| return false if yield x }
+    true
+  end
```
Comment (BrittanyBlake) [inline enumerables.rb:77]: - [ ] When no block or argument is given, `#my_none?` should return true only if none of the collection members is true.

```ruby
[nil, false, nil, false].none? === [nil, false, nil, false].my_none? => false
```

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    y = 0
+    if block_given?
+      arr.my_each { |x| y += 1 if yield x }
+    elsif !block_given? && arg.nil?
+      y = arr.length
+    else
+      y = arr.my_select { |x| x == arg }.length
+    end
+    y
+  end
```
Comment (BrittanyBlake) [inline enumerables.rb]: - [ ] When a range is passed to`#my_count`, it should return the number of items in enum through enumeration.

```ruby
range = Range.new(5, 50)
p range.my_count => error
```
* Yours currently returns a `NoMethodError`: `undefined method 'length' for 5..50:Range`

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+  def my_map(arg = nil)
+    index = 0
+    n_array = []
+    while index < to_a.length
+      n_array[index] = arg.nil? ? yield(to_a[index]) : !yield(to_a[index])
+      index += 1
+    end
+    n_array
+  end
```
Comment (BrittanyBlake) [inline enumerables.rb]: - [ ] When no block is given, `#my_map` should return an enumerator.

```ruby
array = Array.new(100) { rand(0...9) }
p array.my_map => error
```
*  Yours currently returns a` LocalJumpError`:`no block given (yield)`

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
### Status: Changes Required ♻️ 
****
### Hi @gasb150 and @gango-anan! 👋 
You have done a great job implementing the changes requested by the previous TSE! 👍 You have also done a wonderful job with this difficult project so far! 👏 👏 👏  There are just a few issues that need to be resolved before you can move on: ⬇️ 


****
### Please make the required changes and resubmit for a new code review once you are ready 🙏

Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

**Review by ebeagusamuel — APPROVED** [review]
# Project Approved :100: 
Hello team :wave: 
Congratulations on passing all the test cases for this project :clap: :clap: 
You can go ahead and merge this branch.
Good luck with the next project.
slack(@ebeagusamuel)

**Pull request merged**